### PR TITLE
Added Pretty formatter with settings file

### PR DESCRIPTION
**Pull request by IgorProtsenkoOd**
Prettier formatter tool

tabWidth - defines indent size
bracketSpacing - removes extra spaces in { foo: bar }, to become {foo: bar}
semi - terminates every statement with semicolon
singleQuote - replaces all double quotes with single quotes, excluding Json files.
useTabs - defines replacement tabs with spaces

### CODE REVIEW

**Comment by SCdF** [comment]
Interesting PR! I have some questions:
 - Does the lack of code changes elsewhere imply that our current codebase already aligns with this tool completely?
 - Is it required that people use this tool, or is this just another path for people if they don't want to or can't have their editor do the above things for them? (ie my editor deals with tabwidth and useTabs, and it has a jshint plugin that shows me semi and singleQuote live (and is part of our build process).
 - How is this tool supposed to integrate with our workflow? Am I supposed to run something at some point? Is Grunt supposed to?
 - How will we detect when people to not integrate it into their workflow? Would travis run this and fail if it changes things?
 - How stringent is the tool? Will it stop us doing natural manual layouts that are not tool obvious, like:
```js
const MONTH_LENGTHS = {
  JANUARY:   31,
  FEBURARY:  28 // I know, it's just an example
  MARCH:     31,
  APRIL:     30,
  MAY:       31,
  JUNE:      30,
  JULY,      31
  AUGUST:    31,
  SEPTEMBER: 30,
  OCTOBER:   31,
  NOVEMBER:  30,
  DECEMBER:  31
}
```

**Comment by IgorProtsenkoOd** [comment]
1. Not sure that I understood the question. Could you please rephrase that.
2. If I get the question properly, the answer is that tool runs on each commit on precommit hook. So whatever style if was, it will be reformatted and pushed to Git in unified style. Locally files are changed too.
3. If pull is accepted, people have to pull changes and do npm install again.
4. If any developer pulls the code from master and do not run npm install, commit will fail because library for precommit hook is missing. Newcomers get it working by default.
5. In your example it fails,  and you won't be able to commit, as well as it fail in any IDE since syntax is wrong.
Line: `JULY,      31` (expected `JULY: 31,`)
Once July line is fixed, tool will get rid of extra spaces between month names and their values.

**Comment by SCdF** [comment]
> Not sure that I understood the question. Could you please rephrase that.

This PR only adds the tool, there is nothing in this PR that changes the formatting of existing code to comply with this tool. Does that mean that if we run this tool over our entire codebase it thinks all of our code is valid, or are you saying that this is not intended to run over all existing code just new code, or should it run over existing code but hasn't yet?

> If I get the question properly, the answer is that tool runs on each commit on precommit hook. So whatever style if was, it will be reformatted and pushed to Git in unified style. Locally files are changed too.

So it sounds like you'd like it to be required that people use this tool, and specifically that they install a git precommit hook to use it? This answers my next question ("How is this tool supposed to integrate with our workflow?"). I looked into it, and `husky` is a npm module that installs git pre-commit hooks: this is the understanding I was missing.

> Once July line is fixed, tool will get rid of extra spaces between month names and their values.

Yeah sorry, the July bit was a mistake on my part, what I was really interested in was whether it would allow the extra spaces. I looked into it, and you can add comments that tell prettier to ignore something: https://prettier.io/docs/en/ignore.html

From my current understanding we could trial it for a bit, but if we are interested in voting I'd vote against it: I'm very much not a fan of a pre-commit hook that moves code around.

**Comment by IgorProtsenkoOd** [comment]
>This PR only adds the tool, there is nothing in this PR that changes the formatting of existing code to comply with this tool. Does that mean that if we run this tool over our entire codebase it thinks all of our code is valid, or are you saying that this is not intended to run over all existing code just new code, or should it run over existing code but hasn't yet?

If tools is accepted, there could be a separate pull request with all reformatted code. (should be done once, then every commit from anyone with any IDE settings will be the same)

>Yeah sorry, the July bit was a mistake on my part, what I was really interested in was whether it would allow the extra spaces. I looked into it, and you can add comments that tell prettier to ignore something: https://prettier.io/docs/en/ignore.html

Yes, this could be an option, although it's not considered to be a good style for extra spaces in code, as well as extra line breaks (which Prettier also cleans up).

>From my current understanding we could trial it for a bit, but if we are interested in voting I'd vote against it: I'm very much not a fan of a pre-commit hook that moves code around.

but this completely solves the codestyle issue, - everyone uses something he likes the most, but that something requires a lot of setup to match all the style rules in the project.
Even now, having guildline for style, we have double quotes in one module, single quotes in another, bracket spacing in third, no braking spacing in fourth, and so on.
Prettier looks much more preferable option above any other, if we are trying to keep code consistent and keep one codestyle everywhere.

**Comment by SCdF** [comment]
>  it's not considered to be a good style for extra spaces in code, as well as extra line breaks (which Prettier also cleans up).

Not considered to be a good style by whom, and with what reasoning? As a counter thought, I think for example having arrays laid out arbitrarily basing line breaks on character width when you could lay them out logically based on what the data represents to be bad style.

> but this completely solves the codestyle issue

I'm so sorry, I wasn't aware there was one! All code I've seen committed since I got here is basically the same in the core layout (num spaces, where brackets go etc). There are definitely old files that are in a different style, but we left them intentionally to try to keep git blame clean. If we don't like those existing we can run a formatter over them at any point: imo I don't think it really matters that they are different, but my setup allows me to think this way (my text editor just handles it) and it may not be the same for other people.

Can you go into more detail about the codestyle issue as you see it? Or we can just talk at the next standup, I didn't mean for this PR to be a giant conversation :-)

**Comment by garethbowen** [comment]
I agree with this in principle. @IgorProtsenkoOd I'll assign this to Stefan to review while I'm overseas.

@SCdF Having no enforced code style is adding complexity to our onboarding experience - every time we get someone new on there's a few months of PR rejections while we train them to do it our way. This tool has IDE integration as well as command line so it should be really easy to get new hires up to speed.

Some notes:

- The precommit hook should use our existing `grunt precommit` target and prettier be added to that.
- Can we ignore certain blocks with inline comments or so on so prettier will ignore everywhere the layout is important to convey meaning?
- We should discuss at some point exactly what rules we want to enforce.
- We should discuss at some point whether we want to reformat the whole codebase in one go or just reformat as we go.

**Comment by IgorProtsenko** [comment]
@SCdF 
Absolutely, will do.

**Comment by IgorProtsenkoOd** [comment]
@SCdF 
Will do.

**Comment by garethbowen** [comment]
It looks like you can ignore a section with a comment so that solves one of Stefan's concerns: https://prettier.io/docs/en/ignore.html#javascript

**Comment by IgorProtsenkoOd** [comment]
Yes, but in our Todo list is to define what rules to enforce.

File: webapp/src/js/services/db-sync.js [hunk 1]
```diff
@@ -1,19 +1,12 @@
...
-    Session
-  ) {
+  READ_ONLY_TYPES = ['form', 'translations'],
+  READ_ONLY_IDS = ['resources', 'appcache', 'zscore-charts', 'settings'],
+  DDOC_PREFIX = ['_design/'],
+  META_SYNC_FREQUENCY = 30 * 60 * 1000; // 30 minutes
 
+angular
+  .module('inboxServices')
+  .factory('DBSync', function($interval, $log, $q, Auth, DB, Session) {
```
Comment (SCdF) [inline webapp/src/js/services/db-sync.js]: Would we whitelist here?

File: webapp/src/js/services/db-sync.js [hunk 2]
```diff
@@ -32,17 +25,19 @@ angular.module('inboxServices').factory('DBSync',
...
     var readOnlyFilter = function(doc) {
       // don't try to replicate read only docs back to the server
-      return READ_ONLY_TYPES.indexOf(doc.type) === -1 &&
-             READ_ONLY_IDS.indexOf(doc._id) === -1 &&
-             doc._id.indexOf(DDOC_PREFIX) !== 0;
+      return (
+        READ_ONLY_TYPES.indexOf(doc.type) === -1 &&
+        READ_ONLY_IDS.indexOf(doc._id) === -1 &&
+        doc._id.indexOf(DDOC_PREFIX) !== 0
+      );
```
Comment (SCdF) [inline webapp/src/js/services/db-sync.js]: IMO this looks worse (uses two extra vertical lines to make scope clear that was clear already), but it's not the end of the world. I wonder if there is a flag to turn this off?

File: webapp/src/js/services/db-sync.js [hunk 3]
```diff
@@ -32,17 +25,19 @@ angular.module('inboxServices').factory('DBSync',
...
     };
 
     var getOptions = function(direction) {
       var options = {
         live: true,
         retry: true,
         heartbeat: 10000,
-        back_off_function: backOffFunction
+        back_off_function: backOffFunction,
```
Comment (SCdF) [inline webapp/src/js/services/db-sync.js]: I am very cool with this becoming our standard.

**Comment by SCdF** [comment]
@garethbowen I'm interested in trying it on larger parts of the codebase.

FWIW, I'm still not a fan of a pre-commit hook altering code I'm committing[1], but I'm happy to try it! Also, happy to add prettier to ST3 (I'm sure there is a plugin) so hopefully the hook shouldn't *actually* change anything.

[1] Though I guess with `git commit -v` I get to see what it's changed?

File: webapp/src/js/services/db-sync.js (hunk 1 above)
Comment (garethbowen) [reply webapp/src/js/services/db-sync.js]: I don't think I'd whitelist this. I was never 100% happy with how it was. This change makes it a little hard to see which dependencies have been added or removed but only a little.

I think consistency wins here.

File: webapp/src/js/services/db-sync.js (hunk 2 above)
Comment (garethbowen) [reply webapp/src/js/services/db-sync.js]: Not sure. Its tagline is "opinionated code formatter" and documented flags are IMO minimal. I think I'd get used to it and opt for consistency.

File: webapp/src/js/services/db-sync.js (hunk 3 above)
Comment (garethbowen) [reply webapp/src/js/services/db-sync.js]: Yeah, that one is configurable and I turned it on for everywhere that's valid under es5.

**Comment by garethbowen** [comment]
@SCdF If you have time I recommend checking out this branch and formatting some files specifically - the sublime plugin looks easy enough.

My main concerns are doing a code-wide reformat muddies the git history (particularly for blaming), but not doing a code-wide reformat means every commit will include reformatting that's unrelated to the code change. Which do you think is less worse?

**Comment by garethbowen** [comment]
@SCdF Ok, so let's hold off from doing a full reformat at least for now, but otherwise you're ok with this change? Or do you think the precommit hooks should be turned off by default?

**Comment by SCdF** [comment]
@garethbowen let's have them on by default. It's what is most helpful to the team I'd imagine. If it ends up being annoying to many people or otherwise problematic we can change it later.

**Pull request merged**